internal/exec/runner.go

```diff
@@ -0,0 +1,922 @@
+			downloaderOptions,
+			protoc.DownloaderWithCachePath(r.cachePath),
+		)
+	}
+	if r.protocURL != "" {
+		downloaderOptions = append(
+			downloaderOptions,
+			protoc.DownloaderWithProtocURL(r.protocURL),
+		)
+	}
+	return protoc.NewDownloader(config, downloaderOptions...)
+}
+
+func (r *runner) newCompiler(doGen bool, doFileDescriptorSet bool) protoc.Compiler {
+	compilerOptions := []protoc.CompilerOption{
+		protoc.CompilerWithLogger(r.logger),
+	}
+	if r.cachePath != "" {
+		compilerOptions = append(
+			compilerOptions,
+			protoc.CompilerWithCachePath(r.cachePath),
+		)
+	}
+	if r.protocURL != "" {
+		compilerOptions = append(
+			compilerOptions,
+			protoc.CompilerWithProtocURL(r.protocURL),
+		)
+	}
+	if doGen {
+		compilerOptions = append(
+			compilerOptions,
+			protoc.CompilerWithGen(),
+		)
+	}
+	if doFileDescriptorSet {
+		compilerOptions = append(
+			compilerOptions,
+			protoc.CompilerWithFileDescriptorSet(),
+		)
+	}
+	return protoc.NewCompiler(compilerOptions...)
+}
+
+func (r *runner) newLintRunner() lint.Runner {
+	return lint.NewRunner(
+		lint.RunnerWithLogger(r.logger),
+	)
+}
+
+func (r *runner) newTransformer(rewrite bool) format.Transformer {
+	transformerOptions := []format.TransformerOption{format.TransformerWithLogger(r.logger)}
+	if rewrite {
+		transformerOptions = append(transformerOptions, format.TransformerWithRewrite())
+	}
+	return format.NewTransformer(transformerOptions...)
+}
+
+func (r *runner) newGetter() extract.Getter {
+	return extract.NewGetter(
+		extract.GetterWithLogger(r.logger),
+	)
+}
+
+func (r *runner) newReflectHandler() reflect.Handler {
+	return reflect.NewHandler(
+		reflect.HandlerWithLogger(r.logger),
+	)
+}
+
+func (r *runner) newCreateHandler(pkg string) create.Handler {
+	handlerOptions := []create.HandlerOption{create.HandlerWithLogger(r.logger)}
+	if pkg != "" {
+		handlerOptions = append(handlerOptions, create.HandlerWithPackage(pkg))
+	}
+	return create.NewHandler(handlerOptions...)
+}
+
+func (r *runner) newGRPCHandler(
+	headers map[string]string,
+	callTimeout time.Duration,
+	connectTimeout time.Duration,
+	keepaliveTime time.Duration,
+) grpc.Handler {
+	handlerOptions := []grpc.HandlerOption{
+		grpc.HandlerWithLogger(r.logger),
+	}
+	for key, value := range headers {
+		handlerOptions = append(handlerOptions, grpc.HandlerWithHeader(key, value))
+	}
+	if callTimeout != 0 {
+		handlerOptions = append(handlerOptions, grpc.HandlerWithCallTimeout(callTimeout))
+	}
+	if connectTimeout != 0 {
+		handlerOptions = append(handlerOptions, grpc.HandlerWithConnectTimeout(connectTimeout))
+	}
+	if keepaliveTime != 0 {
+		handlerOptions = append(handlerOptions, grpc.HandlerWithKeepaliveTime(keepaliveTime))
+	}
+	return grpc.NewHandler(handlerOptions...)
+}
+
+func (r *runner) getConfig(dirPath string) (settings.Config, error) {
+	return r.configProvider.GetForDir(dirPath)
+}
+
+type meta struct {
+	ProtoSets               []*file.ProtoSet
+	InDirModeSingleFilename string
```

Comment:
can you comment this? It wasn't clear as to what this was indicating initially.

Might be worth separating this as two args: `bool` and `string`, but that's optional

Reply:
Separating leads to more bugs here potentially, think of this as an atomic thing, no inconsistent state where `dirMode == false && singleFilename == non-empty`. Commented.